src/main/java/com/google/flourbot/execution/MacroExecutionModuleImplementation.java

```diff
@@ -56,23 +55,76 @@ private String getMacroName(String message, String threadId) {
       throw new IllegalStateException(e);
     }
 
-    this.threadMacroMap.put(threadId, macroName);
+    threadMacroMap.put(threadId, macroName);
     return macroName;
   }
 
+  public String getReplyText(String[] words, String message, String threadId, String roomId, String messageSenderEmail, String helpMessage) throws IOException, GeneralSecurityException{
+        String macroName;
 
-  public String execute(String userEmail, String message, String threadId) throws IOException, GeneralSecurityException {
-    //Check if the message is a help message
-    String[] words = message.split(" ", 2);
-    if (words[1].equalsIgnoreCase("help")) {
-        return "To use your macro, please type \"@MacroBot MacroName <your message>\". If your macro has already been used in a room's thread, you may omit the MacroName and can simply write \"@MacroBot <your message>\". ";
-    }
-    
-    String macroName = this.getMacroName(message, threadId);
+        // Check that a message was written.
+        if (words.length <= 1) {
+            return "You must type a message when you message me. Please type \"@MacroBot /help\" for more instructions.";
+        }    
+
+        // Reply to the help message.
+        if (words[1].equalsIgnoreCase("/help")) {  
+            return helpMessage;
+        }
+
+        // Get the macro name from the message. 
+        if (words[1].equalsIgnoreCase("/share")) {
+            String modifiedMessage = removeShareFromMessage(words);
+            macroName = getMacroName(modifiedMessage, threadId);
+        }  
+        else {
+            macroName = getMacroName(message, threadId);
+        }
+
+        // If the macro has already been shared to this room, execute the macro. 
+        if (roomToMacro.containsKey(roomId)) {
+            Map<String, String> macroToCreator = roomToMacro.get(roomId);
+            String macroCreatorEmail = macroToCreator.get(macroName);
+            return execute(messageSenderEmail, macroCreatorEmail, message, threadId, macroName);
+        }
+        // If the macro has not been shared to the room, query the database for the macro.
+        else {
+            Optional<Macro> optionalMacro = entityModule.getMacro(messageSenderEmail, macroName);
+            if (!optionalMacro.isPresent()) {
+                return "You do not own/have access to this macro.";
+            }
+            else {
+                // Share the macro if the creator writes "/share".
+                if (words[1].equalsIgnoreCase("/share")) {
```

Comment:
if(isShareCommand)

If we have a clumsy method of checking a condition and there is nothing we can do about it, it is better to do it once and keep the state in the variable.

Reply:
Sounds good, I've added the varaible.